src/bluez/adapter/peripheral.rs

```diff
@@ -11,410 +11,404 @@
 //
 // Copyright (c) 2014 The Rust Project Developers
 
+use dbus::{
+    arg::{RefArg, Variant},
+    blocking::{stdintf::org_freedesktop_dbus::PropertiesPropertiesChanged, Proxy, SyncConnection},
+    channel::Token,
+    message::{Message, SignalArgs},
+    Path,
+};
+
+use bytes::BufMut;
+
 use crate::{
     api::{
-        AddressType, BDAddr, Callback, Central, CharPropFlags, Characteristic, CommandCallback,
-        NotificationHandler, Peripheral as ApiPeripheral, PeripheralProperties, RequestCallback,
-        UUID, UUID::B16,
+        AdapterManager, AddressType, BDAddr, CentralEvent, CharPropFlags, Characteristic,
+        CommandCallback, NotificationHandler, Peripheral as ApiPeripheral, PeripheralProperties,
+        RequestCallback, ValueNotification, UUID,
     },
     bluez::{
-        adapter::acl_stream::ACLStream,
-        adapter::ConnectedAdapter,
-        constants::*,
-        protocol::{att, hci, hci::ACLData},
-        util::handle_error,
+        bluez_dbus::device::OrgBluezDevice1, AttributeType, Handle, BLUEZ_DEST, DEFAULT_TIMEOUT,
     },
     Error, Result,
 };
 
 use std::{
-    collections::{BTreeSet, VecDeque},
+    collections::{BTreeSet, HashMap},
     fmt::{self, Debug, Display, Formatter},
-    mem::size_of,
-    sync::{
-        atomic::Ordering,
-        mpsc::{self, channel, Receiver, Sender},
-        Arc, Condvar, Mutex, RwLock,
-    },
-    time::Duration,
+    sync::{Arc, Condvar, Mutex},
+    time::Instant,
 };
 
-use bytes::{BufMut, BytesMut};
-use libc;
-
-#[derive(Copy, Debug)]
-#[repr(C)]
-pub struct SockaddrL2 {
-    l2_family: libc::sa_family_t,
-    l2_psm: u16,
-    l2_bdaddr: BDAddr,
-    l2_cid: u16,
-    l2_bdaddr_type: u8,
-}
-impl Clone for SockaddrL2 {
-    fn clone(&self) -> Self {
-        *self
-    }
-}
-
-#[derive(Copy, Debug, Default)]
-#[repr(C)]
-struct L2CapOptions {
-    omtu: u16,
-    imtu: u16,
-    flush_to: u16,
-    mode: u8,
-    fcs: u8,
-    max_tx: u8,
-    txwin_size: u16,
-}
-impl Clone for L2CapOptions {
-    fn clone(&self) -> Self {
-        *self
-    }
+#[derive(Clone, Debug, PartialEq, PartialOrd)]
+enum PeripheralState {
+    NotConnected,
+    Connected,
+    ServicesResolved,
 }
 
 #[derive(Clone)]
 pub struct Peripheral {
-    c_adapter: ConnectedAdapter,
+    adapter: AdapterManager<Self>,
+    connection: Arc<SyncConnection>,
+    path: String,
     address: BDAddr,
     properties: Arc<Mutex<PeripheralProperties>>,
     characteristics: Arc<Mutex<BTreeSet<Characteristic>>>,
-    stream: Arc<RwLock<Option<ACLStream>>>,
-    connection_tx: Arc<Mutex<Sender<u16>>>,
-    connection_rx: Arc<Mutex<Receiver<u16>>>,
-    message_queue: Arc<Mutex<VecDeque<ACLData>>>,
+    attributes_map: Arc<Mutex<HashMap<u16, (String, Handle, Characteristic)>>>,
```

Comment:
If you wanted to lose the Mutex here, you could switch to DashMap. I use it heavily up in Buttplug, works great. That said, with everything else so lock heavy here, that's really just the lightest of nits and could be a followup.